From 4159ca91cc06ec06c61f2dc0803570b17bee3978 Mon Sep 17 00:00:00 2001
From: Dan Kephart <dan.kephart@lairdtech.com>
Date: Mon, 6 Jan 2014 10:40:49 -0500
Subject: [PATCH] Sending radio mode before/after disconnect cmd and regdomain
 events

BZ4478
Firmware could default to unwanted radio mode when sending a disconnect
request when the firmware isn't connected.  Also, regdomains can change
radio modes so send an event after seeing a regdomain event.
---
 drivers/net/wireless/ath/ath6kl/wmi.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/wireless/ath/ath6kl/wmi.c b/drivers/net/wireless/ath/ath6kl/wmi.c
index 8e5b379..9427836 100644
--- a/drivers/net/wireless/ath/ath6kl/wmi.c
+++ b/drivers/net/wireless/ath/ath6kl/wmi.c
@@ -4027,6 +4027,8 @@ static int ath6kl_wmi_proc_events_vif(struct wmi *wmi, u16 if_idx, u16 cmd_id,
 
 static int ath6kl_wmi_proc_events(struct wmi *wmi, struct sk_buff *skb)
 {
+	struct ath6kl *ar = wmi->parent_dev;
+	struct ath6kl_vif *vif;
 	struct wmi_cmd_hdr *cmd;
 	int ret = 0;
 	u32 len;
@@ -4037,6 +4039,7 @@ static int ath6kl_wmi_proc_events(struct wmi *wmi, struct sk_buff *skb)
 	cmd = (struct wmi_cmd_hdr *) skb->data;
 	id = le16_to_cpu(cmd->cmd_id);
 	if_idx = le16_to_cpu(cmd->info1) & WMI_CMD_HDR_IF_ID_MASK;
+	vif = ath6kl_get_vif_by_index(ar, if_idx);
 
 	skb_pull(skb, sizeof(struct wmi_cmd_hdr));
 	datap = skb->data;
@@ -4071,6 +4074,8 @@ static int ath6kl_wmi_proc_events(struct wmi *wmi, struct sk_buff *skb)
 		break;
 	case WMI_REGDOMAIN_EVENTID:
 		ath6kl_dbg(ATH6KL_DBG_WMI, "WMI_REGDOMAIN_EVENTID\n");
+		if(vif && vif->sme_state != SME_CONNECTED)
+			ath6kl_wmi_send_radio_mode(wmi, if_idx);
 		ath6kl_wmi_regdomain_event(wmi, datap, len);
 		break;
 	case WMI_PSTREAM_TIMEOUT_EVENTID:
-- 
1.8.1.2

